packages/expo-in-app-purchases/android/src/main/java/expo/modules/inapppurchases/BillingManager.java

```diff
@@ -253,27 +275,44 @@ public int getBillingClientResponseCode() {
   }
 
   /**
-   * Query purchases across various use cases and deliver the result in a formalized way through
-   * a listener
+   * Query both in app purchases and subscriptions and deliver the result in a formalized way 
+   * through a listener
    */
   public void queryPurchases(final Promise promise) {
     Runnable queryToExecute = new Runnable() {
       @Override
       public void run() {
-        PurchasesResult purchasesResult = mBillingClient.queryPurchases(SkuType.INAPP);
-
-        // If there are subscriptions supported, we add subscription rows as well
-        if (purchasesResult != null && areSubscriptionsSupported()) {
-          PurchasesResult subscriptionResult
-            = mBillingClient.queryPurchases(SkuType.SUBS);
-
-          if (subscriptionResult != null && subscriptionResult.getResponseCode() == BillingResponseCode.OK) {
-            purchasesResult.getPurchasesList().addAll(
-              subscriptionResult.getPurchasesList());
+        final List<String> ALL_QUERIES = Arrays.asList(SkuType.INAPP, SkuType.SUBS);
+        List<Purchase> purchases = new ArrayList<>();
+        List<String> completedQueries = new ArrayList<>();
+
+        mBillingClient.queryPurchasesAsync(SkuType.INAPP, new PurchasesResponseListener() {
+          @Override
+          public void onQueryPurchasesResponse(@NonNull BillingResult billingResult, @NonNull List<Purchase> inAppPurchases) {
+            if (billingResult.getResponseCode() == BillingResponseCode.OK) {
+              purchases.addAll(inAppPurchases);
+            }
+            completedQueries.add(SkuType.INAPP);
+            if (completedQueries.containsAll(ALL_QUERIES) || !areSubscriptionsSupported()) {
+              onQueryPurchasesFinished(billingResult, purchases, promise);
```

Comment:
The onQueryPurchasesFinished calls look potentially buggy since they pass in the BillingResult for one query but not both.

1. Is passing in the last BillingResult correct? Or do we want to aggregate data from the 1-2 BillingResults?
2. Is the old code buggy?

Reply:
The only part of BillingResult we're ever interested in is the response code. Although this is _most likely_ equivalent for both queries (INAPP and SUBS), it's potentially buggy since maybe we could get a `SERVICE_DISCONNECTED` or `SERVICE_TIMEOUT` in one query and not the other. Unlikely, but possible, and if that were to happen the current code would return the purchases it _did_ get, and ignore the error (or vice versa- ignore the good purchases, and return an error).

So, yes the old code could have hidden response codes we should've surfaced, and we should probably make it respond consistently in all cases, so maybe the best thing to do is aggregate from the BillingResults:

- If all are OK, pass any one of them (doesn't matter which, all we're after is responseCode)
- If one is NOT OK, pass that one
- if both are NOT OK, pass either one

Reply:
I think we should expose partial failures, not hide them. Exactly what you proposed.